### DISCUSSION TITLE
Multi-threaded Domain Decomp

### DISCUSSION THREAD

**Discussion by roger82374** [OpenRadioss/OpenRadioss#1987] [Q&A]
Hi I am trying to speed up domain decomposition in starter using multi-threaded:

I am following this:

https://2021.help.altair.com/2021/hwsolvers/rad/topics/solvers/rad/rad_user_guide_intro_c.htm

```
Use a Different Number of Cores

The Starter can use multiple cores via SMP parallelization.

For extremely large models that will run using a large number of SPMD domains, the Starter will run faster if multiple SMP cores are utilized for domain decomposition and to create the restart files. When using the Compute Console (ACC) or included script, the Starter and Engine must be ran separately using the -onestep option. For example, assuming a very large 3 million element model that will run on 120 cores on a compute server with 12 cores per CPU.

The Starter will use 12 cores to calculate the 120 SPMD domains and create the restart files.

radioss -nt 12 -np 120 -onestep model_0000.rad

The Engine will use 120 cores using SPMD parallelization.

radioss -nt 1 -np 120 -onestep model_0001.rad
```

-onestep doesn't seem to be an option & using -nt 8 doesn't seem to change anything

Thank you, any suggestions?

**Comment by MarianBulla** [comment]
Hi,
thank you for your interest on OpenRadioss.
regarding your question, if you want to use 120 cores for your model, and you have 10 cpus with 12 cores in each cpu, you can go with
-np 10 and nt 12 for the engine.
- The starter should be used with -np 10
- The engine should run using -np 10 and -nt 12
Which gives you 10x12 = 120 cores.

For more options and info, you can check: https://www.open-mpi.org/
Or the oneAPI HPC from intel, if you are using intel MPI.

The "-onestep" option runs only the starter. This is for testing/debugging purpose of your model. So, you don't need it, if your model is ok and working well.

Hope it helps you ?

Thanks and best regards,
Marian

**Comment by roger82374** [comment]
Thanks Marian,

Perhaps I didn't understand the guide above. So if I wish to run a model just in parallel on 1*16 CPU  (-np 16 and -nt 1), I cannot run the starter on multiple threads to speed up the decomposition?

Thank you

**Reply by MarianBulla** [reply]
Yes, you can run the STARTER with several threads.
Just use the starter with -np 16 (so, it will decompose your model for 16 CPUs) and -nt xxx to get this job done faster.

Hope it helps you ?

Thansk and best
Marian

**Reply by roger82374** [reply]
Hi Marian,

I don't think that function is working for me, I can see that the machine only uses one cpu and the time does not improve

Thank you

**Reply by MarianBulla** [reply]
Hi,
ok, I will test it on my side and provide you my findings.
Will come back with some findings...later

But I can test it only on 1CPU with multiple threads, if that will help you ?

Thansk and best
Marian

**Reply by MarianBulla** [reply]
Hi, 
please have a loop, how the command works:

![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/e1c53ac5-4ff0-4728-b1bd-0ee8f570b21f)

![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/cf6146e9-94bf-469f-b23d-0a22df9976e0)

Hope it helps you ?

Thanks and best regards,
Marian

**Reply by roguic** [reply]
Hi @roger82374 ,

The OpenMP parrallelization of the starter is limited to 1 part : the creation of the restart files. The other parts of the starter are sequential (so without any OpenMP parrallelization). If your model is a small model, then the restart file creation will be quick and increase the number of OpenMP threads will not improve the performance of the starter. If your model is a large model, then OpenMP parrallelization will reduce the elapsed time.

Best,
Romain

**Reply by roger82374** [reply]
Thanks, that's what I see with the Taurus example. I think I was looking for more of a difference which was my confusion. Thank you for your help!

![Pasted image](https://github.com/OpenRadioss/OpenRadioss/assets/148344562/8a6d4410-1fc4-4860-a540-bff63854d60b)
![Pasted image 1](https://github.com/OpenRadioss/OpenRadioss/assets/148344562/4941d127-c7cb-4558-bc36-a6287a4d8ab4)

**Reply by roguic** [reply]
Hi @roger82374 ,

I saw in your log that you used 56 threads. 
If your processor has several sockets, the best choice for the OpenMP thread number is the number of physical threads of 1 socket (for some affinity question). 
Example : 
1 processor with 2 sockets, 8 physical threads per socket --> Best choice for OpenMP threads = 8
You can use more than 8 OpenMP threads -> the performance may be improved but the performance improvement is not guaranteed.

Best,
Romain

**Reply by roger82374** [reply]
Hi @roguic,

This is just for the starter I assume, so in my case I have 2*28 core Xeon (total 112 hyperthreaded). So in my case a better number would be 28 threads to use? I assume this is something to do with latency between NUMA nodes?

Thank you

**Reply by PaulAltair** [reply]
In fact running the starter on many threads is not usually helpful, and even on engine, I wouldn't normally run more than -nt 8, for your hardware I would compare -nt 1, -np 56 and -nt 2 -np 28, there is a rule of thumb that the model decomposition benefits begin to fall away at less than 10k nodes per processor, so if your model is smaller than 280000 nodes, benefits of running spmd may be limited when running on all nodes

**Reply by PaulAltair** [reply]
Also, in your original post, some of the options you refer to, e.g. -onestep are only relevant to Altair compute console submission. For OpenRadioss if you are running from commandline or using the gui submission, you may use different -nt options in starter and engine submission, but -np should be the same for both

**Reply by roguic** [reply]
Hi @roger82374 ,

@PaulAltair is right : 10k nodes per processor is the limit. If your model uses complex material laws or expensive elements, maybe you can try 5000 nodes per processor (for the taurus 1M element model, 10k nodes per processor is a good compromise).

A short remainder for the engine : if you use OpenMPI, you need to map the mpi tasks and bind the OpenMP threads 
for X mpi tasks and Y OpenMP threads per mpi tasks :
export OMP_NUM_THREADS=Y
export OMP_STACKSIZE=400M
mpirun -np X --map-by socket:PE=$OMP_NUM_THREADS --bind-to core $PATH_TO_EXEC/./engine_linux64_gf_ompi -i *1.rad

Best,
Romain

**Reply by roger82374** [reply]
Thank you both that is very helpful advice
